### ISSUE TITLE
MOX-159: Add local runtime observability for warm/cold transitions, active sessions, queue depth, memory footprint, and backend health changes

### ISSUE THREAD

**Issue by AtlantisPleb** [OpenAgentsInc/openagents#3199]
Roadmap ID: MOX-159.

Source:
- `crates/mox/docs/ROADMAP.md`
- Section: Epic C: Behavioral contract and serving policy

## Goal
Add local runtime observability for warm/cold transitions, active sessions, queue depth, memory footprint, and backend health changes

## Why this matters now
The desktop cutover will be hard to debug without explicit runtime-state observability.

## Scope
`mox-serve`, `mox-provider`, `mox-runtime`

## Design constraints
- Respect `docs/OWNERSHIP.md` boundaries.
- Keep capability, fallback, and evidence surfaces explicit and machine-checkable.
- Do not hide backend, quantization, lifecycle, or degraded-state behavior behind silent defaults.
- Keep reusable engine behavior in `crates/mox/*` rather than app-specific product orchestration.

## Acceptance criteria
- The work is implemented within the scoped crates or owner listed in the roadmap.
- Resulting behavior is testable and documented enough for downstream cutover or backend issues.
- Capability, conformance, or evidence implications are explicit where this issue changes them.

## Dependencies
- Follow the ordered dependency plan in `crates/mox/docs/ROADMAP.md`.

**Comment by AtlantisPleb** [comment]
Landed on main in a7b73314f (). I used Ollama's scheduler/runner lifecycle as the primary reference, especially the ref-counted warm/idle/unload and backend-health seams in  and the runner health endpoint, but kept Mox library-first instead of daemon/log oriented.

What landed:
- reusable runtime observability types in  for backend health observations, bounded recent transition logs, and a serializable local-runtime snapshot
- lifecycle transition tracking in  for cold load, first-request warm transition, unload/expiry, active sessions, active requests, memory footprint, and backend health state
- a first-class  surface on the managed in-process runtime plus provider-facing observability envelope serialization/tests

Verification:
- 
running 26 tests
test parity::tests::default_policy_keeps_cpu_as_reference_and_quantized_budget_looser ... ok
test parity::tests::embedding_comparison_uses_budget_and_cosine_similarity ... ok
test parity::tests::generation_budget_is_exact_when_seeded_and_semantic_when_unseeded ... ok
test tests::amd_backend_model_serializes_mode_topology_risk_and_recovery ... ok
test tests::backend_health_tracker_records_only_real_changes ... ok
test tests::backend_extension_support_serializes_stably ... ok
test parity::tests::logits_comparison_allows_one_rank_drift_for_quantized_paths ... ok
test tests::backend_selection_helpers_capture_direct_and_fallback_truth ... ok
test tests::backend_runtime_resources_serialize_stably ... ok
test tests::kv_cache_state_and_growth_serialize_stably ... ok
test tests::kv_page_layout_reports_page_and_byte_geometry ... ok
test tests::buffer_handles_can_distinguish_quantized_storage_from_dequantized_fallback ... ok
test tests::loaded_model_residency_tracks_keepalive_and_request_activity ... ok
test tests::mock_runtime_reports_device_and_executes_plan ... ok
test tests::model_admission_can_evict_oldest_idle_model_to_fit_budget ... ok
test tests::model_admission_refuses_when_only_active_models_block_the_budget ... ok
test tests::local_runtime_observability_serializes_stably ... ok
test tests::paged_tensor_storage_plan_serializes_byte_window_and_page_counts ... ok
test tests::prefix_cache_identity_and_policy_serialize_stably ... ok
test tests::quantization_support_surfaces_storage_path_and_pending_execution_truth ... ok
test tests::runtime_model_storage_truth_distinguishes_paged_blobs_from_copies ... ok
test tests::runtime_transition_log_retains_latest_events ... ok
test tests::residency_policy_and_memory_plan_serialize_stably ... ok
test tests::sampling_penalties_honor_the_bounded_lookback_window ... ok
test tests::sampling_policy_serializes_supported_generation_controls ... ok
test tests::seeded_token_sampler_replays_draws ... ok

test result: ok. 26 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s


running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
- 
running 51 tests
test conformance::tests::generate_case_builder_uses_real_qwen2_fixture ... ok
test conformance::tests::compare_show_ignores_candidate_only_mox_facts ... ok
test conformance::tests::conformance_suite_accepts_matching_prompt_render_candidate ... ok
test conformance::tests::conformance_suite_fails_on_embed_drift ... ok
test conformance::tests::local_ollama_catalog_subject_reports_missing_model ... ok
test conformance::tests::performance_gate_refuses_cutover_when_metrics_are_missing ... ok
test conformance::tests::report_writer_emits_pretty_json ... ok
test conformance::tests::local_ollama_catalog_subject_show_reports_adapter_policy_facts ... ok
test conformance::tests::performance_gate_accepts_candidate_within_generation_budget ... ok
test tests::cpu_reference_text_generation_rejects_unknown_session ... ok
test conformance::tests::local_ollama_catalog_subject_lists_and_shows_models ... ok
test tests::cpu_reference_text_generation_refuses_prompt_context_overflow ... ok
test tests::embedding_output_rejects_non_finite_values ... ok
test tests::embedding_request_json_is_stable ... ok
test tests::embedding_response_round_trips ... ok
test tests::generation_response_round_trips ... ok
test tests::generation_request_json_is_stable ... ok
test tests::generation_sampling_options_round_trip ... ok
test tests::generation_sessions_isolate_and_reset_kv_cache ... ok
test tests::generation_sessions_reject_descriptor_drift_even_when_model_id_matches ... ok
test tests::golden_fixture_corpus_is_visible_from_serve ... ok
test tests::golden_prompt_render_cases_are_reusable_from_serve ... ok
test tests::golden_prompt_window_cases_are_reusable_from_serve ... ok
test tests::model_registry_can_evict_oldest_idle_model_to_admit_new_candidate ... ok
test tests::cpu_reference_generation_stream_disconnect_returns_terminal ... ok
test tests::model_registry_refuses_candidate_when_host_budget_would_be_exceeded ... ok
test tests::model_registry_records_lifecycle_transitions ... ok
test tests::cpu_reference_text_generation_can_truncate_oldest_prompt_tokens ... ok
test tests::model_registry_tracks_keepalive_order_and_idle_expiry ... ok
test tests::model_registry_zero_keepalive_unloads_when_idle ... ok
test tests::model_registry_tracks_active_generation_models ... ok
test tests::penalties_shift_token_selection ... ok
test tests::cpu_reference_observability_reports_sessions_memory_and_transitions ... ok
test tests::cpu_reference_generation_stream_emits_chunks_then_terminal ... ok
test tests::paged_kv_cache_tracks_growth_refill_and_refusal ... ok
test tests::seeded_sampling_is_replayable ... ok
test tests::generation_stream_returns_error_before_first_chunk_when_prompt_is_empty ... ok
test tests::cpu_reference_text_generation_is_deterministic ... ok
test tests::cpu_reference_text_generation_session_budget_counts_existing_context ... ok
test tests::cpu_reference_generation_stream_cancellation_discards_uncommitted_session_output ... ok
test tests::smoke_embeddings_service_is_deterministic ... ok
test tests::generation_stream_reports_runtime_failure_after_stream_start ... ok
test tests::shared_prefix_store_reports_hit_miss_and_rebuilt ... ok
test tests::cpu_reference_text_generation_truncates_stop_sequences ... ok
test tests::cpu_reference_text_generation_updates_loaded_model_residency ... ok
test tests::cpu_reference_text_generation_reports_cold_then_warm_provenance ... ok
test tests::cpu_reference_text_generation_replays_seeded_sampling_options ... ok
test tests::cpu_reference_text_generation_reuses_and_resets_sessions ... ok
test tests::mox_local_runtime_forwards_catalog_lifecycle_generation_and_embeddings ... ok
test conformance::tests::ollama_http_subject_normalizes_live_http_responses ... ok
test tests::cpu_reference_text_generation_reports_prefix_hits_and_bypasses ... ok

test result: ok. 51 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s


running 1 test
test metal_model_backed_embeddings_parity_reports_explicit_offline_state ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s


running 1 test
test metal_model_backed_embeddings_flow_returns_response_capability_and_receipt_or_explicit_unavailability ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s


running 5 tests
test model_backed_embeddings_service_reports_missing_artifact ... ok
test model_backed_embeddings_empty_batch_returns_empty_success ... ok
test model_backed_embeddings_reject_wrong_product ... ok
test model_backed_embeddings_reject_reference_descriptor_without_fallback ... ok
test model_backed_embeddings_flow_returns_response_capability_and_receipt ... ok

test result: ok. 5 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s


running 4 tests
test model_backed_text_generation_service_reports_missing_artifact ... ok
test model_backed_text_generation_rejects_reference_descriptor_without_fallback ... ok
test model_backed_text_generation_rejects_unknown_session ... ok
test model_backed_text_generation_flow_returns_response_capability_and_receipt ... ok

test result: ok. 4 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s


running 2 tests
test smoke_embeddings_empty_batch_returns_empty_success ... ok
test smoke_embeddings_flow_returns_response_capability_and_receipt ... ok

test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s


running 1 test
test cpu_reference_text_generation_flow_returns_response_and_receipt ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s


running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
- 
running 17 tests
test tests::amd_kfd_capability_reports_mode_topology_and_recovery ... ok
test tests::amd_receipt_carries_execution_context ... ok
test tests::amd_userspace_capability_reports_disabled_risk_posture ... ok
test tests::embedding_request_digest_changes_when_output_dimensions_change ... ok
test tests::capability_envelope_preserves_backend_runtime_resources ... ok
test tests::failed_receipt_carries_reason ... ok
test tests::fallback_capability_reports_requested_metal_but_effective_cpu ... ok
test tests::capability_envelope_json_is_stable ... ok
test tests::execution_receipt_round_trips ... ok
test tests::generation_request_digest_changes_when_options_change ... ok
test tests::readiness_helper_sets_ready_status ... ok
test tests::local_runtime_observability_envelope_serializes_stably ... ok
test tests::request_digests_are_deterministic ... ok
test tests::request_digests_change_when_weight_identity_changes ... ok
test tests::streaming_terminal_receipt_preserves_partial_cancellation_output ... ok
test tests::text_generation_capability_json_is_stable ... ok
test tests::text_generation_receipt_round_trips ... ok

test result: ok. 17 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s


running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
- Ownership boundary check passed.
- 

Note: Touched-file clippy gate skipped: no changed Rust files. is still blocked by pre-existing unrelated  clippy debt in  ( at line 1085 and  at line 1302).
